File: src/promouters/schemas/expense_plans.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field

from promouters.models.enums import ExpenseApprovalDecision, ExpensePlanStatus


class ExpensePlanItemCreate(BaseModel):
    sequence: int = Field(default=1, ge=1)
    name: str = Field(min_length=1, max_length=255)
    category: str | None = Field(default=None, max_length=100)
    quantity: Decimal = Field(default=Decimal("1"), gt=0, decimal_places=2, max_digits=12)
    unit_price: Decimal = Field(ge=0, decimal_places=2, max_digits=12)
    note: str | None = None


class ExpensePlanItemRead(BaseModel):
    id: UUID
    expense_plan_id: UUID
    sequence: int
    name: str
    category: str | None
    quantity: Decimal
    unit_price: Decimal
    amount: Decimal
    note: str | None


class ExpensePlanCreate(BaseModel):
    title: str = Field(default="План расходов", min_length=1, max_length=255)
    branch_id: UUID
    period_start: date
    period_end: date
    currency: str = Field(default="RUB", min_length=3, max_length=3)
    comment: str | None = None
    items: list[ExpensePlanItemCreate] = Field(default_factory=list)


class ExpensePlanUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    period_start: date | None = None
    period_end: date | None = None
    currency: str | None = Field(default=None, min_length=3, max_length=3)
    comment: str | None = None
    items: list[ExpensePlanItemCreate] | None = None


class ExpenseApprovalRead(BaseModel):
    id: UUID
    expense_plan_id: UUID
    approver_id: UUID
    approver_name: str
    decision: str
    comment: str | None
    decided_at: datetime | None
    created_at: datetime


class ExpenseApprovalDecisionRequest(BaseModel):
    decision: ExpenseApprovalDecision
    comment: str | None = None


class ExpensePlanRead(BaseModel):
    id: UUID
    branch_id: UUID
    branch_name: str
    created_by_id: UUID
    created_by_name: str
    title: str
    period_start: date
    period_end: date
    total_amount: Decimal
    currency: str
    status: str
    comment: str | None
    submitted_at: datetime | None
    approved_at: datetime | None
    items: list[ExpensePlanItemRead]
    approvals: list[ExpenseApprovalRead]
    created_at: datetime
    updated_at: datetime

File: frontend/src/shared/api/expense-plans.ts
import { apiRequest } from './http';

export type ExpensePlanStatus =
  | 'draft'
  | 'submitted'
  | 'approved'
  | 'rejected'
  | 'cancelled';

export type ExpenseDecision = 'pending' | 'approved' | 'rejected' | 'needs_revision';

export type ExpensePlanItem = {
  id: string;
  expense_plan_id: string;
  sequence: number;
  name: string;
  category: string | null;
  quantity: string | number;
  unit_price: string | number;
  amount: string | number;
  note: string | null;
};

export type ExpenseApproval = {
  id: string;
  expense_plan_id: string;
  approver_id: string;
  approver_name: string;
  decision: ExpenseDecision;
  comment: string | null;
  decided_at: string | null;
  created_at: string;
};

export type ExpensePlanRecord = {
  id: string;
  branch_id: string;
  branch_name: string;
  created_by_id: string;
  created_by_name: string;
  title: string;
  period_start: string;
  period_end: string;
  total_amount: string | number;
  currency: string;
  status: ExpensePlanStatus;
  comment: string | null;
  submitted_at: string | null;
  approved_at: string | null;
  items: ExpensePlanItem[];
  approvals: ExpenseApproval[];
  created_at: string;
  updated_at: string;
};

export type ExpensePlanItemPayload = {
  sequence?: number;
  name: string;
  category?: string | null;
  quantity: number;
  unit_price: number;
  note?: string | null;
};

export type ExpensePlanCreatePayload = {
  title?: string;
  branch_id: string;
  period_start: string;
  period_end: string;
  currency?: string;
  comment?: string | null;
  items?: ExpensePlanItemPayload[];
};

export type ExpensePlanDecisionPayload = {
  decision: ExpenseDecision;
  comment?: string | null;
};

export async function fetchExpensePlans(accessToken: string) {
  return apiRequest<ExpensePlanRecord[]>('/expense-plans', {
    method: 'GET',
    accessToken,
  });
}

export async function fetchExpensePlan(accessToken: string, planId: string) {
  return apiRequest<ExpensePlanRecord>(`/expense-plans/${planId}`, {
    method: 'GET',
    accessToken,
  });
}

export async function createExpensePlan(
  accessToken: string,
  payload: ExpensePlanCreatePayload,
) {
  return apiRequest<ExpensePlanRecord>('/expense-plans', {
    method: 'POST',
    accessToken,
    body: JSON.stringify(payload),
  });
}

export async function submitExpensePlan(accessToken: string, planId: string) {
  return apiRequest<ExpensePlanRecord>(`/expense-plans/${planId}/submit`, {
    method: 'POST',
    accessToken,
  });
}

export async function decideExpensePlan(
  accessToken: string,
  planId: string,
  payload: ExpensePlanDecisionPayload,
) {
  return apiRequest<ExpensePlanRecord>(`/expense-plans/${planId}/decision`, {
    method: 'POST',
    accessToken,
    body: JSON.stringify(payload),
  });
}
